File: SymmetricEncryptionAPI/SymmetricEncryptionAPI/Helpers/AesEncryption.cs
using System.Security.Cryptography;
using System.Text;

namespace SymmetricEncryptionAPI.Helpers
{
    public class AesEncryption
    {
        private readonly byte[] _key;

        public AesEncryption(string key)
        {
            // Generate a 256-bit key from the given string using SHA256
            using (var sha256 = SHA256.Create())
            {
                _key = sha256.ComputeHash(Encoding.UTF8.GetBytes(key));
            }
        }

        public string Encrypt(string plainText)
        {
            // Generate a unique IV for each encryption
            using (var aes = Aes.Create())
            {
                aes.Key = _key;
                aes.GenerateIV(); // Create a random IV
                var iv = aes.IV;

                using (var encryptor = aes.CreateEncryptor(aes.Key, iv))
                using (var msEncrypt = new MemoryStream())
                {
                    // Prepend IV to the memory stream
                    msEncrypt.Write(iv, 0, iv.Length);

                    using (var csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write))
                    using (var swEncrypt = new StreamWriter(csEncrypt))
                    {
                        swEncrypt.Write(plainText);
                    }
                    return Convert.ToBase64String(msEncrypt.ToArray());
                }
            }
        }

        public string Decrypt(string cipherText)
        {
            try
            {
                // Decode the base64-encoded cipherText
                var fullCipher = Convert.FromBase64String(cipherText);

                // Extract IV from the first 16 bytes of fullCipher
                var iv = new byte[16];
                Array.Copy(fullCipher, 0, iv, 0, iv.Length);

                // Extract the actual ciphertext (without the IV)
                var actualCipherText = new byte[fullCipher.Length - iv.Length];
                Array.Copy(fullCipher, iv.Length, actualCipherText, 0, actualCipherText.Length);

                using (var aes = Aes.Create())
                {
                    aes.Key = _key;
                    aes.IV = iv;
                    aes.Padding = PaddingMode.PKCS7;

                    using (var decryptor = aes.CreateDecryptor(aes.Key, aes.IV))
                    using (var msDecrypt = new MemoryStream(actualCipherText))
                    using (var csDecrypt = new CryptoStream(msDecrypt, decryptor, CryptoStreamMode.Read))
                    using (var srDecrypt = new StreamReader(csDecrypt))
                    {
                        // Read and return the decrypted plaintext
                        return srDecrypt.ReadToEnd();
                    }
                }
            }
            catch (CryptographicException ex)
            {
                // Log the specific padding error for debugging
                Console.WriteLine("Decryption failed: Padding is invalid and cannot be removed.");
                throw new InvalidOperationException("Decryption failed. Ensure encryption and decryption configurations match.", ex);
            }
        }

    }
}

File: symmetric-encryption-react/src/utils/aesEncryption/AesEncryption.js
import CryptoJS from 'crypto-js';

// Secret key
export const key = "YourEncryptionKey";

// Function to encrypt plaintext
export const encrypt = (plainText, key) => {
    // Generate a 256-bit key using SHA256 hash
    const hashedKey = CryptoJS.SHA256(key).toString(CryptoJS.enc.Hex);

    // Generate a random 16-byte IV
    const iv = CryptoJS.lib.WordArray.random(16);

    // Encrypt with AES
    const encrypted = CryptoJS.AES.encrypt(plainText, CryptoJS.enc.Hex.parse(hashedKey), {
        iv: iv,
        mode: CryptoJS.mode.CBC,
        padding: CryptoJS.pad.Pkcs7,
    });

    // Combine IV and ciphertext, then encode in Base64
    return CryptoJS.enc.Base64.stringify(iv.concat(encrypted.ciphertext));
};


// Function to decrypt ciphertext
export const decrypt = (cipherText, key) => {
    // Decode the base64-encoded cipherText
    const fullCipher = CryptoJS.enc.Base64.parse(cipherText);

    // Extract IV from the first 16 bytes (128 bits / 8 = 16 bytes)
    const iv = CryptoJS.lib.WordArray.create(fullCipher.words.slice(0, 4), 16);

    // Extract the actual ciphertext without the IV
    const actualCipherText = CryptoJS.lib.WordArray.create(
        fullCipher.words.slice(4),
        fullCipher.sigBytes - 16
    );

    // Generate the hashed key as a WordArray
    const hashedKey = CryptoJS.SHA256(key);

    // Decrypt with AES using CBC mode and PKCS7 padding
    const decrypted = CryptoJS.AES.decrypt(
        { ciphertext: actualCipherText },
        hashedKey,
        {
            iv: iv,
            mode: CryptoJS.mode.CBC,
            padding: CryptoJS.pad.Pkcs7,
        }
    );

    // Convert decrypted data to UTF-8 string
    const data = decrypted.toString(CryptoJS.enc.Utf8);

    return data;
};